platform/unix/rxsubcom.1
.TH RXSUBCOM 1 "April 2021" "Version 5.0.0"
.SH NAME
\fBrxsubcom\fP \- Open Object Rexx Subcommand Administrator
.SH SYNTAX
.B rxsubcom
.RI "REGISTER " "environment sharedLibrary function"

.B rxsubcom
.RI "QUERY " "sharedLibrary " [ function ]

.B rxsubcom
.RI "DROP " "sharedLibrary " [ function ]

.B rxsubcom
.RI "LOAD " "sharedLibrary " [ function ]

.SH DESCRIPTION
The
.B rxsubcom
command registers, drops, and queries Rexx subcommand handlers. A Rexx program can use
.B rxsubcom
to register dynamic-link library subcommand handlers. Once the subcommand handler is registered,
the Rexx program can send commands to the subcommand handler with the Rexx ADDRESS instruction.

.SH "SEE ALSO"
.BR rxqueue (1)

.SH EXAMPLES
.TP
rxsubcom register mysub libmysub.so StartMySub
Registers the
.B mysub
environment in the
.B libmysub.so
shared library

.TP
rxsubcom drop libmysub.so
Drops the subcommand environment provided by the shared library
.B libmysub.so

.SH LICENSE
This program and the accompanying materials are made available under
the terms of the Common Public License v1.0 which accompanies this
distribution. A copy is also available at the following address:
https://www.oorexx.org/license.html


CHANGES
            Fixes in Open Object Rexx 5.0.0

The following is a list of fixes, enhancements, and other relevant changes
to ooRexx since the 4.2.0 release.  The numbers for each item can be used to
look up the item in the appropriate tracker in the ooRexx project on
SourceForge, i.e. the Bugs tracker, the Feature Requests tracker, etc.


  Overview / Major Changes
  ------------------------

  Keyword instructions
     new DO WITH INDEX .. ITEM .. OVER ..          [feature-requests:#158]
     new FOR iterator for DO WITH and DO OVER      [feature-requests:#605]
     new SELECT CASE                               [feature-requests:#600]
     new USE LOCAL ..                              [feature-requests:#654]
     added call type REQUIRES to PARSE SOURCE      [feature-requests:#622]

  Built-in functions
     added CONDITION("RESET")                      [feature-requests:#682]
     added DATATYPE("Internal") format             [feature-requests:#711]
     added DATE("ISO") format                      [feature-requests:#407]
     added XRANGE() named character sequences      [feature-requests:#639]

  Added notation
     new [namespace]:[class|routine] notation      [feature-requests:#464]
     new (item1, item2, ..) Array notation         [feature-requests:#613]
     new "REXX" package                            [feature-requests:#602]

  Directives
     new ::ANNOTATE ATTRIBUTE/CLASS/CONSTANT/METHOD/PACKAGE/ROUTINE [feature-requests:#620]
     added ABSTRACT subkeyword to ::CLASS          [feature-requests:#631]
     added DELEGATE subkeyword to ::ATTRIBUTE and ::METHOD [feature-requests:#649]
     value on ::CONSTANT now optional              [feature-requests:#389]
     new ::OPTIONS NOVALUE CONDITION/ERROR         [feature-requests:#568]
                                                   [feature-requests:#394]
     new ::OPTIONS PROLOG/NOPROLOG                 [feature-requests:#621]
     new ::RESOURCE                                [feature-requests:#545]
                                                   [feature-requests:#61]
     added NAMESPACE option to ::REQUIRES          [feature-requests:#464]

  Changes to classes and methods
     new class AlarmNotification                   [feature-requests:#644]
     new class json                                [feature-requests:#625]
     new class MessageNotification                 [feature-requests:#640]
     new class RexxInfo                            [feature-requests:#618]
     new class StringTable                         [feature-requests:#615]
     new class Ticker                              [feature-requests:#653]
     new class Validate                            [feature-requests:#707]

     new methods added to Alarm class
       canceled/cancelled                          [feature-requests:#652]
       scheduledTime                               [feature-requests:#646]
       attachment                                  [feature-requests:#645]

     new methods added to Class class
       annotation, annotations                     [feature-requests:#620]
       defineMethods                               [feature-requests:#647]
       isMetaclass                                 [feature-requests:#629]
       package                                     [feature-requests:#596]

     new method added to Collection classes
       of                                          [feature-requests:#616]
     changed existing method of Collection classes
       size argument added to method new           [feature-requests:#604]
       appendAll, putAll, fill, empty              [feature-requests:#670]

     new method added to Directory class
       removeEntry                                 [feature-requests:#607]

     new methods added to Message class
       halt                                        [feature-requests:#642]
       hasResult                                   [feature-requests:#606]
       sendWith, startWith (send/start changed)    [feature-requests:#643]
       wait                                        [feature-requests:#641]
     changes
       allow messages triggered more than once     [feature-requests:#245]

     new methods added to Method class
       annotation, annotations                     [feature-requests:#620]
       isAbstract, isConstant, and isAttribute     [feature-requests:#632]
       scope                                       [feature-requests:#636]

     changed existing methods of Method class
       optional "context" argument added to newFile [feature-requests:#686]

     new methods added to MutableBuffer class
       [], []=                                     [feature-requests:#614]
       (caseless)Contains, (caseless)ContainsWord  [feature-requests:#451]
       (caseless)startsWith, (caseless)endsWith    [feature-requests:#651]
       setText                                     [feature-requests:#599]
     changed existing methods of MutableBuffer class
       append allows multiple strings              [feature-requests:#619]
       match/matchChar position restriction removed [feature-requests:#650]

     new method added to Object class
       isNil                                       [feature-requests:#696]

     new methods added to Package class
       annotation, annotations                     [feature-requests:#620]
       findNamespace, namespaces                   [feature-requests:#464]
       findProgram                                 [feature-requests:#648]
       findPublicClass, findPublicRoutine          [feature-requests:#630]
       local                                       [feature-requests:#708]
       prolog                                      [feature-requests:#638]
       resource, resources                         [feature-requests:#545]
                                                   [feature-requests:#61]
     changed existing methods of Package class
       optional "context" argument added to new    [feature-requests:#633]

utilities/rxsubcom/platform/windows/rxsubcom.cpp
/*----------------------------------------------------------------------------*/
/*                                                                            */
/* Copyright (c) 1995, 2004 IBM Corporation. All rights reserved.             */
/* Copyright (c) 2005-2020 Rexx Language Association. All rights reserved.    */
/*                                                                            */
/* This program and the accompanying materials are made available under       */
/* the terms of the Common Public License v1.0 which accompanies this         */
/* distribution. A copy is also available at the following address:           */
/* https://www.oorexx.org/license.html                                        */
/*                                                                            */
/* Redistribution and use in source and binary forms, with or                 */
/* without modification, are permitted provided that the following            */
/* conditions are met:                                                        */
/*                                                                            */
/* Redistributions of source code must retain the above copyright             */
/* notice, this list of conditions and the following disclaimer.              */
/* Redistributions in binary form must reproduce the above copyright          */
/* notice, this list of conditions and the following disclaimer in            */
/* the documentation and/or other materials provided with the distribution.   */
/*                                                                            */
/* Neither the name of Rexx Language Association nor the names                */
/* of its contributors may be used to endorse or promote products             */
/* derived from this software without specific prior written permission.      */
/*                                                                            */
/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS        */
/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT          */
/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS          */
/* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT   */
/* OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,      */
/* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED   */
/* TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,        */
/* OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY     */
/* OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING    */
/* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS         */
/* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.               */
/*                                                                            */
/*----------------------------------------------------------------------------*/
/*********************************************************************/
/*                                                                   */
/* Descriptive Name:   Windows Command Line Program for Subcommand   */
/*                     Interface.                                    */
/*                                                                   */
/*********************************************************************/
#include <string.h>
#include <stdio.h>

#include "rexx.h"
#include "RexxInternalApis.h"
#include "RexxErrorCodes.h"

#define CASE(x) if(!_stricmp(x, argv[1]))

void parmerr(int);

int __cdecl main(int argc, char *argv[], char *envp[])
{
    int args = argc - 2;               // number of RXSUBCOM [ACTION] args

    if (args < 0)                      // no RXSUBCOM argument specified
    {
        // The RXSUBCOM parameters are incorrect
        parmerr(Error_RXSUBC_general);
    }

    CASE("REGISTER")
    {
        if (args != 3)                 // REGISTER envname dllname procname
        {
            // The RXSUBCOM REGISTER parameters are incorrect
            parmerr(Error_RXSUBC_register);
        }
        return RexxRegisterSubcomDll(argv[2], argv[3], argv[4], NULL, RXSUBCOM_DROPPABLE);
    }

    CASE("QUERY")
    {
        if (args < 1 || args > 2)      // QUERY envname [dllname]
        {
            // The RXSUBCOM QUERY parameters are incorrect
            parmerr(Error_RXSUBC_query);
        }
        unsigned short flags;
        return RexxQuerySubcom(argv[2], args == 1 ? NULL : argv[3], &flags, NULL);
    }

    CASE("DROP")
    {
        if (args < 1 || args > 2)      // DROP envname [dllname]
        {
            // The RXSUBCOM DROP parameters are incorrect
            parmerr(Error_RXSUBC_drop);
        }
        return RexxDeregisterSubcom(argv[2], args == 1 ? NULL : argv[3]);
    }

    CASE("LOAD")
    {
        if (args < 1 || args > 2)      // LOAD envname [dllname]
        {
            // The RXSUBCOM LOAD parameters are incorrect
            parmerr(Error_RXSUBC_load);
        }
        return RexxLoadSubcom(argv[2], args == 1 ? NULL : argv[3]);
    }

    // The RXSUBCOM parameters are incorrect
    parmerr(Error_RXSUBC_general);
    return 0;                          // this will never be reached
}

void parmerr(int msgid )
{
    // retrieve the message from the central catalog and print it
    const char *message = RexxGetErrorMessage(msgid);
    printf("%s\n", message);

    exit(-1);
}
